Enable zoom/attribute dependent stream rendering

mmcfarland opened the pull request:
#### Overview
For both NHD Plus and DRB streams, configure the cartographic styling
to limit streams of a high Stream Order (scale is 10-0, large to small)
to be visible at low zoom levels (high altitude).  As the zoom level
increases, more streams of lower stream order are shown and the stroke
width of the rendered line corresponds to the relative stream order.
The effect should be a gradual increase in detail of streams as the
user zooms in, and streams of a higher Stream Order should appear
larger than those of a smaller order.

The intent isn't to make the stream order attribute apparent from
viewing the stream, but to give a relative indication of size and limit
the number of streams shown when zoomed out.

Since the DRB and NHD streams vary considerably in density of streams
at every zoom level, some of the sizes were adjusted to accommodate a
reasonable appearance for both datasets while keeping the width of the
stream the same, per stream order, per dataset.

Connects #1409 
Connects #1411 

#### Testing
* Restart your tile and app server to ensure that the latest styling changes and zoom configurations are present
  * `vagrant ssh tiler -c "sudo service mmw-tiler restart"`
  * `vagrant ssh app -c "sudo service mmw-app restart"`
* Starting at zoom 7, enable the National Stream overlay (this is not fast, since it's doing a large geographic query, but will be cached in stage/prod)
* Zoom in, noticing that more "smaller" streams appear the closer you get
* Do the same in the DRB, starting at zoom 10

#### Code evaluation
It will be helpful to understand that in CartoCSS `[][]` is a logical `AND` and `[], []` is a logical `OR`.

This may not be the final set of zoom/attribute breaks, I'm sure the client will have feedback for a desired distribution.

mmcfarland commented:
#### DRB Streams
![drb_streams](https://cloud.githubusercontent.com/assets/1014341/17895587/96194684-691b-11e6-94c8-61d4f7da3b1f.gif)

mmcfarland commented:
#### NHD Streams
![nhd_streams_sm](https://cloud.githubusercontent.com/assets/1014341/17895726/0590adcc-691c-11e6-97cf-07ec0feb4f91.gif)

mmcfarland commented:
Updated with a commit that renames both steam layers.  This is to avoid the CloudFront/s3 caching we have in place in staging.  Each PR that adjusts these styles can rename the layer so that changes are visible immediately with waiting for caches to expire or needing to purge them on staging/production.

rajadain commented:
I restarted the `mmw-tiler`service, reloaded the `tiler` VM, then reprovisioned it, but I'm still getting a 400 on all the requests:

![image](https://cloud.githubusercontent.com/assets/1430060/17907088/23f89ce4-6948-11e6-845e-d9709fabf614.png)

rajadain commented:
Here's a response detail:

![image](https://cloud.githubusercontent.com/assets/1430060/17907131/5bc41220-6948-11e6-9256-0883bb3e4584.png)

rajadain commented:
Ah, looking at the tiler log I see the following:

```
[TILE RENDER ERROR] -- 400: Error: Postgis Plugin: ERROR:  column "stream_order" does not exist
LINE 1: SELECT ST_AsBinary("geom") AS geom,"stream_order" FROM drb_s...
                                           ^

in executeQuery Full sql was: 'SELECT ST_AsBinary("geom") AS geom,"stream_order" FROM drb_streams_50 WHERE "geom" && ST_SetSRID('BOX3D(-75.14923095703125 39.94133068417753,-75.09979248046875 39.97922477515792)'::box3d, 4326)'
```

I probably need to update my local streams datasets. Doing so now.

mmcfarland commented:
Oh sorry - yeah, that work didn't coincide with this PR so it slipped my mind, but this does require an updated NHD stream set.  The old DRB already had a stream order attribute, but i renamed it in a data update to match NHD so that we could use the same rendering rules for both.  You'll need to update both.

rajadain commented:
Consider renaming `drb_streams` here as well: https://github.com/WikiWatershed/model-my-watershed/blob/c4891a56401cce469c634e1330a2ba5c0511c492/src/mmw/js/src/draw/views.js#L81

rajadain commented:
If we're moving to a convention of always using new IDs for the caching layer, can we excise the `purge_tile_cache` command here https://github.com/WikiWatershed/model-my-watershed/blob/77982cc669568de3c8c0658f5440de0e76619e62/scripts/aws/setupdb.sh#L68-L72 ? The [specified](https://github.com/WikiWatershed/model-my-watershed/blob/77982cc669568de3c8c0658f5440de0e76619e62/scripts/aws/setupdb.sh#L101) [paths](https://github.com/WikiWatershed/model-my-watershed/blob/77982cc669568de3c8c0658f5440de0e76619e62/scripts/aws/setupdb.sh#L110) are no longer valid.

rajadain commented:
+1 tested, they look much nicer now, and I didn't have to wait very long for them to load either.

`setupdb` with `nhdflowline` was taking forever within the `app` VM, so I had to manually do the steps by saving the 1.2 GB file to disk, then unzipping it to ~3.8 GB, and finally running `psql` on that, which was relatively quicker. I think my `app` VM may not have had enough RAM to be able to stream that via pipes very quickly.

mmcfarland commented:
Good catch with the missed rename, addressed in 03ce620.  

Also added a commit that improves rendering performance at very low zoom levels to satisfy a request at showing high stream_orders at a high altitude view.  Slightly tweaked the rendering rules, but the testing should now include viewing both stream sets at high levels.

![nhd](https://cloud.githubusercontent.com/assets/1014341/17916352/94a8e376-6980-11e6-91ce-5cee17d8ce38.png)
![drb](https://cloud.githubusercontent.com/assets/1014341/17916353/94ae81aa-6980-11e6-8731-1387973843ea.png)

mmcfarland commented:
Re: the cache purge paths, I was planning on changing the names back to their original names in a subsequent PR, after however many number of revisions of style tweaks.  I'm going to keep the old purge names for now and manually clean up the s3 buckets for these new ones unless this cache bypass ends up working out better than the cache purge.

rajadain commented:
Cool.

It's really interesting seeing the visualizations at the national scale. I don't think I ever had a sense of the major rivers in the country until now.

mmcfarland commented on src/tiler/server.js:
```diff
@@ -124,7 +147,14 @@ var config = {
             var tableId = req.params.tableId,
                 tableName = tables[tableId];
 
+
```
I'll take out this extra line.

rajadain commented:
+1 tested the fixups. Works well. Great job!
